oracle_fdw.c

```diff
@@ -732,6 +734,50 @@ _PG_init(void)
 	on_proc_exit(&exitHook, PointerGetDatum(NULL));
 }
 
+/*
+ * Deparse LIMIT clause translated for ORACLE into FETCH FIRST N ROWS ONLY
+ * If OFFSET is set then current offset value is added to the LIMIT value
+ * so that it will give the Oracle optimizer the right clue.
+ */
+char *
+deparseLimit(PlannerInfo *root, struct OracleFdwState *fdwState, RelOptInfo *baserel)
+{
+	StringInfoData limit_clause;
+
+	initStringInfo(&limit_clause);
+
+	if (root->parse->limitCount != NULL && !root->parse->groupClause)
+	{
+		if (IsA(root->parse->limitCount, Const) && !((Const *) root->parse->limitCount)->constisnull)
+		{
+			char *limit_val;
+			char *offset_val;
+
+			appendStringInfoString(&limit_clause, " FETCH FIRST ");
+			limit_val = deparseExpr(
+						fdwState->session, baserel,
+						(Expr *) root->parse->limitCount,
+						fdwState->oraTable,
+						&(fdwState->params)
+					);
+			if (root->parse->limitOffset != NULL)
```

Comment:
You check that the `LIMIT` is not NULL above; I think you should do the same for `OFFSET`.

Reply:
I do not understand, line 763 show a test on offset is not null. Actually there is limitCount for the LIMIT clause and limitOffset for the OFFSET clause.

Reply:
You check if there is a `LIMIT` clause at all, but you don't check if `IsA(root->parse->limitOffset, Const)` or if it has `->constisnull`.

Reply:
ok got it, fixed in commit 7720b7d as well as removing call to deparseExpr.